modules/authglobus/lib/Auth/Source/Globus.php

```diff
@@ -0,0 +1,213 @@
+
+    /**
+     * Constructor for this authentication source.
+     *
+     * @param array $info  Information about this authentication source.
+     * @param array $config  Configuration.
+     */
+    public function __construct($info, $config) {
+        // Call the parent constructor first, as required by the interface
+        parent::__construct($info, $config);
+        
+        $this->key = $config['key'];
+        $this->secret = $config['secret'];
+        $this->scope = $config['scope'];
+        $this->response_type = $config['response_type'];
+        $this->redirect_uri= $config['redirect_uri'];
+    }
+
+    /**
+     * Flattens an associative array, derived from LinkedIn module's flatten
+     *
+     * @param array $array
+     * @param string $prefix
+     *
+     * @return array the array with the new concatenated keys
+     */
+    protected function flatten($array, $prefix = '')
+    {
+        $newArr = array();
+        foreach ($array as $key => $value) {
+            if (is_array($value)) {
+                $newArr = $newArr + $this->flatten($value, $prefix . $key . '.');
+            } else {
+                $newArr[$prefix . $key] = $value;
+            }
+        }
+        return $newArr;
+    }
+
+    /**
+     * Exchange an authorization code for Globus Tokens.
+     *
+     * @param array $contents The contents of the POST request. REF: https://docs.globus.org/api/auth/reference/#authorization_code_grant_preferred
+     * @param string $signature Your Client ID + Secret from the Globus app you've registered.
+     */
+    private function getGlobusTokens($contents, $signature) {
+        $curl = curl_init();
+
+        curl_setopt($curl, CURLOPT_USERAGENT, 'SSPHP Globus'); //umad Globus
+        curl_setopt($curl, CURLOPT_USERPWD, $signature);
+        curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
+        curl_setopt($curl, CURLOPT_FOLLOWLOCATION, true);
+        curl_setopt($curl, CURLOPT_RETURNTRANSFER, 1);
+
+        curl_setopt($curl, CURLOPT_SSL_VERIFYHOST, false);
+        curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);
+
+        curl_setopt($curl, CURLOPT_URL, API_ENDPT . '/oauth2/token');
+        curl_setopt($curl, CURLOPT_POSTFIELDS, OauthUtil::build_http_query($contents));
+
+        $newArr = curl_exec($curl);
+        $outInfo = curl_getinfo($curl);
+
+        return $newArr;
+    }
+
+    /**
+     * Retrieve user info/identity affiliations from the Globus API
+     *
+     * @param string $url The url + endpoints we want to hit
+     * @param array $token The bearer who will make the request
+     */
+    private function getGlobusInfo($url, $token) {
+        $header = 'Authorization: Bearer ' . $token;
+        $options = array(
+            'http' => array(
+                'method' => 'GET',
+                'header' => $header
+            )
+        );
+
+        $context = stream_context_create($options);
+        $info = file_get_contents($url, false, $context);
```

Comment:
what happens if you get an error back?

we should be checking for errors.

Which can be done in a curl method...